File: ios/PretextDiagnostics.swift
import Foundation

internal func buildParagraphLayoutDiagnostics(
    paragraph: NativePreparedParagraph,
    corpus: NativePreparedCorpus,
    request: NativeLayoutRequest,
    lineLayouts: [NativeLineLayout]
) -> ParagraphLayoutDiagnostics {
    let breakTable = buildParagraphBreakTable(paragraph: paragraph, lineLayouts: lineLayouts)
    let boundaryMap = buildParagraphBoundaryMap(
        paragraph: paragraph,
        lineLayouts: lineLayouts,
        breakTable: breakTable
    )
    let driftKinds = collectDriftKinds(
        paragraph: paragraph,
        corpus: corpus,
        request: request,
        lineLayouts: lineLayouts,
        boundaryMap: boundaryMap
    )
    return ParagraphLayoutDiagnostics(
        normalizedRequest: buildPublicLayoutRequest(request),
        ruleLayer: ruleLayerPretextNative,
        canvasPixelParityTarget: false,
        textDirection: corpus.baseStyle.textDirection,
        layoutEngine: lineLayouts.first?.layoutEngine ?? layoutEngineIosCoreText,
        heightMetricSource: heightMetricSourcePlatformTextEngineMetrics,
        fallbackReason: lineLayouts.compactMap { $0.fallbackReason }.first,
        driftKinds: driftKinds,
        heightMetricDrivers: heightMetricDrivers,
        breakTable: breakTable,
        boundaryMap: boundaryMap,
        complexShapeCounters: buildComplexShapeCounters(
            paragraph: paragraph,
            boundaryMap: boundaryMap
        ),
        lineDiagnostics: lineLayouts.map { line in
            ParagraphLineDiagnostics(
                textStart: Double(line.textStartUTF16),
                textEnd: Double(line.textEndUTF16),
                textDirection: corpus.baseStyle.textDirection,
                layoutEngine: line.layoutEngine,
                heightMetricSource: heightMetricSourcePlatformTextEngineMetrics,
                fallbackReason: line.fallbackReason,
                driftKinds: collectLineDriftKinds(line),
                clusterViolationOffsets: collectLineClusterViolationOffsets(
                    line: line,
                    boundaryMap: boundaryMap,
                    atomicSpans: paragraph.atomicSpans
                ).map(Double.init)
            )
        }
    )
}

private func buildPublicLayoutRequest(_ request: NativeLayoutRequest) -> ParagraphLayoutRequest {
    ParagraphLayoutRequest(
        width: request.width,
        left: request.left,
        whiteSpace: request.whiteSpace,
        wordBreak: request.wordBreak,
        shapeSlices: request.shapeSlices.map { slice in
            ParagraphShapeSlice(
                top: slice.top,
                height: slice.height,
                left: slice.left,
                width: slice.width
            )
        }
    )
}

private func buildParagraphBreakTable(
    paragraph: NativePreparedParagraph,
    lineLayouts: [NativeLineLayout]
) -> ParagraphBreakTable {
    let textLength = paragraph.text.length
    let nativeSoftBreaks = lineLayouts.dropLast().compactMap { line -> ParagraphBreakOpportunity? in
        let offset = line.textEndUTF16
        guard offset > 0, offset < textLength else {
            return nil
        }
        guard paragraph.text.character(at: offset) != 0x0A else {
            return nil
        }
        return ParagraphBreakOpportunity(
            offset: Double(offset),
            kind: breakKindNativeSoft,
            source: line.layoutEngine
        )
    }

    return ParagraphBreakTable(
        hardBreaks: collectHardBreaks(paragraph.text),
        nativeSoftBreaks: nativeSoftBreaks,
        graphemeBoundaries: collectGraphemeBoundaries(paragraph.text),
        atomicSpans: paragraph.atomicSpans.map { span in
            ParagraphAtomicSpan(
                textStart: Double(span.startUTF16),
                textEnd: Double(span.endUTF16),
                source: span.source
            )
        }
    )
}

private func buildParagraphBoundaryMap(
    paragraph: NativePreparedParagraph,
    lineLayouts: [NativeLineLayout],
    breakTable: ParagraphBreakTable
) -> ParagraphBoundaryMap {
    let graphemeBoundaries = breakTable.graphemeBoundaries.map(Int.init).sorted()
    let runBoundaries = collectRunBoundaries(paragraph: paragraph)
    let atomicSpanBoundaries = Array(
        Set(paragraph.atomicSpans.flatMap { span in
            [span.startUTF16, span.endUTF16]
        })
    ).sorted()
    let clusterViolations = collectParagraphClusterViolationOffsets(
        lineLayouts: lineLayouts,
        graphemeBoundarySet: Set(graphemeBoundaries),
        atomicSpans: paragraph.atomicSpans
    )

    return ParagraphBoundaryMap(
        utf16Length: Double(paragraph.text.length),
        graphemeBoundaries: graphemeBoundaries.map(Double.init),
        runBoundaries: runBoundaries.map(Double.init),
        hardBreaks: breakTable.hardBreaks.map(\.offset),
        nativeSoftBreaks: breakTable.nativeSoftBreaks.map(\.offset),
        atomicSpanBoundaries: atomicSpanBoundaries.map(Double.init),
        clusterViolationOffsets: clusterViolations.map(Double.init)
    )
}

private func collectHardBreaks(_ text: NSString) -> [ParagraphBreakOpportunity] {
    var breaks: [ParagraphBreakOpportunity] = []
    var cursor = 0
    while cursor < text.length {
        let range = text.rangeOfComposedCharacterSequence(at: cursor)
        if text.substring(with: range) == newlineToken {
            breaks.append(
                ParagraphBreakOpportunity(
                    offset: Double(range.location + range.length),
                    kind: breakKindHard,
                    source: breakSourceNewline
                )
            )
        }
        cursor = range.location + range.length
    }
    return breaks
}

private func collectGraphemeBoundaries(_ text: NSString) -> [Double] {
    var boundaries: [Double] = [0]
    var cursor = 0
    while cursor < text.length {
        let range = text.rangeOfComposedCharacterSequence(at: cursor)
        cursor = range.location + range.length
        boundaries.append(Double(cursor))
    }
    return boundaries
}

private func collectRunBoundaries(paragraph: NativePreparedParagraph) -> [Int] {
    Array(
        Set(
            [0, paragraph.text.length] + paragraph.runs.flatMap { run in
                [run.startUTF16, run.endUTF16]
            }
        )
    )
    .map { min(max($0, 0), paragraph.text.length) }
    .sorted()
}

private func collectParagraphClusterViolationOffsets(
    lineLayouts: [NativeLineLayout],
    graphemeBoundarySet: Set<Int>,
    atomicSpans: [NativeAtomicSpan]
) -> [Int] {
    Array(
        Set(
            lineLayouts.flatMap { line in
                collectLineClusterViolationOffsets(
                    line: line,
                    graphemeBoundarySet: graphemeBoundarySet,
                    atomicSpans: atomicSpans
                )
            }
        )
    ).sorted()
}

private func collectLineClusterViolationOffsets(
    line: NativeLineLayout,
    boundaryMap: ParagraphBoundaryMap,
    atomicSpans: [NativeAtomicSpan]
) -> [Int] {
    collectLineClusterViolationOffsets(
        line: line,
        graphemeBoundarySet: Set(boundaryMap.graphemeBoundaries.map(Int.init)),
        atomicSpans: atomicSpans
    )
}

private func collectLineClusterViolationOffsets(
    line: NativeLineLayout,
    graphemeBoundarySet: Set<Int>,
    atomicSpans: [NativeAtomicSpan]
) -> [Int] {
    var violations = Set<Int>()
    if !graphemeBoundarySet.contains(line.textStartUTF16) {
        violations.insert(line.textStartUTF16)
    }
    if !graphemeBoundarySet.contains(line.textEndUTF16) {
        violations.insert(line.textEndUTF16)
    }
    for span in atomicSpans {
        if line.textStartUTF16 > span.startUTF16 && line.textStartUTF16 < span.endUTF16 {
            violations.insert(line.textStartUTF16)
        }
        if line.textEndUTF16 > span.startUTF16 && line.textEndUTF16 < span.endUTF16 {
            violations.insert(line.textEndUTF16)
        }
    }
    return Array(violations).sorted()
}

private func buildComplexShapeCounters(
    paragraph: NativePreparedParagraph,
    boundaryMap: ParagraphBoundaryMap
) -> ParagraphComplexShapeCounters {
    let boundaries = boundaryMap.graphemeBoundaries.map(Int.init)
    return ParagraphComplexShapeCounters(
        bidiRunCount: Double(countBidiRuns(paragraph.text)),
        emojiClusterCount: Double(
            countClusters(text: paragraph.text, boundaries: boundaries, predicate: containsEmoji)
        ),
        complexClusterCount: Double(countComplexClusters(text: paragraph.text, boundaries: boundaries)),
        clusterViolationCount: Double(boundaryMap.clusterViolationOffsets.count)
    )
}

private func countClusters(
    text: NSString,
    boundaries: [Int],
    predicate: (NSString) -> Bool
) -> Int {
    zip(boundaries, boundaries.dropFirst()).filter { pair in
        let (start, end) = pair
        return end > start &&
            predicate(text.substring(with: NSRange(location: start, length: end - start)) as NSString)
    }.count
}

private func countComplexClusters(text: NSString, boundaries: [Int]) -> Int {
    zip(boundaries, boundaries.dropFirst()).filter { pair in
        let (start, end) = pair
        guard end > start else {
            return false
        }
        let cluster = text.substring(with: NSRange(location: start, length: end - start))
        let scalarCount = cluster.unicodeScalars.count
        return scalarCount > 1 || cluster.unicodeScalars.contains { scalar in
            isCombiningMark(scalar)
                || isVariationSelector(scalar)
                || isEmojiModifier(scalar)
                || isRegionalIndicator(scalar)
                || isIndicVirama(scalar)
                || scalar.value == zeroWidthJoiner
        }
    }.count
}

private func countBidiRuns(_ text: NSString) -> Int {
    var runs = 0
    var previousDirection: Int?
    for scalar in (text as String).unicodeScalars {
        let direction: Int?
        if isRtlScalar(scalar) {
            direction = 1
        } else if isLtrScalar(scalar) {
            direction = 0
        } else {
            direction = nil
        }
        if let direction, direction != previousDirection {
            runs += 1
            previousDirection = direction
        }
    }
    return runs
}

private func collectLineDriftKinds(_ line: NativeLineLayout) -> [String] {
    var driftKinds: [String] = []
    if line.layoutEngine != layoutEngineIosCoreText {
        driftKinds.append(driftEngine)
    }
    if line.fallbackReason != nil {
        driftKinds.append(driftHeightMetric)
    }
    return driftKinds
}

private func collectDriftKinds(
    paragraph: NativePreparedParagraph,
    corpus: NativePreparedCorpus,
    request: NativeLayoutRequest,
    lineLayouts: [NativeLineLayout],
    boundaryMap: ParagraphBoundaryMap
) -> [String] {
    var driftKinds: [String] = []
    func appendDrift(_ driftKind: String) {
        if !driftKinds.contains(driftKind) {
            driftKinds.append(driftKind)
        }
    }

    if lineLayouts.contains(where: { $0.layoutEngine != layoutEngineIosCoreText || $0.fallbackReason != nil }) {
        appendDrift(driftEngine)
    }
    if lineLayouts.contains(where: { $0.fallbackReason != nil }) {
        appendDrift(driftHeightMetric)
    }
    if request.whiteSpace != whiteSpaceNormal
        || request.wordBreak != wordBreakNormal
        || !request.shapeSlices.isEmpty
        || !paragraph.atomicSpans.isEmpty {
        appendDrift(driftAlgorithmRule)
        appendDrift(driftLineBreakStrategy)
    }
    if !corpus.baseStyle.includeFontPadding {
        appendDrift(driftPadding)
    }
    if paragraph.runs.contains(where: { !$0.style.locale.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) {
        appendDrift(driftLocaleMetric)
    }
    if containsPotentialFallbackGlyph(paragraph.text) {
        appendDrift(driftFallbackFont)
    }
    if containsEmoji(paragraph.text) {
        appendDrift(driftEmojiMetric)
    }
    if !boundaryMap.clusterViolationOffsets.isEmpty {
        appendDrift(driftClusterBoundary)
    }
    return driftKinds
}

private func containsPotentialFallbackGlyph(_ text: NSString) -> Bool {
    (text as String).unicodeScalars.contains { scalar in
        scalar.value > 0x02AF
    }
}

private func containsEmoji(_ text: NSString) -> Bool {
    (text as String).unicodeScalars.contains { scalar in
        let value = scalar.value
        return (0x1F000...0x1FAFF).contains(value)
            || (0x2600...0x27BF).contains(value)
            || value == 0xFE0F
    }
}

File: android/src/main/java/com/margelo/nitro/pretext/PretextDiagnostics.kt
package com.margelo.nitro.pretext

internal fun buildParagraphLayoutDiagnostics(
  paragraph: NativePreparedParagraph,
  corpus: NativePreparedCorpus,
  request: NativeLayoutRequest,
  lineLayouts: List<NativeLineLayout>,
): ParagraphLayoutDiagnostics {
  val canonicalLayoutEngine = LAYOUT_ENGINE_ANDROID_STATIC_LAYOUT_COMPAT
  val layoutEngine = lineLayouts.firstOrNull()?.layoutEngine ?: canonicalLayoutEngine
  val fallbackReason = lineLayouts.firstNotNullOfOrNull { it.fallbackReason }
  val breakTable = buildParagraphBreakTable(paragraph, lineLayouts)
  val boundaryMap = buildParagraphBoundaryMap(paragraph, lineLayouts, breakTable)
  val complexShapeCounters = buildComplexShapeCounters(paragraph, boundaryMap)
  return ParagraphLayoutDiagnostics(
    normalizedRequest = buildPublicLayoutRequest(request),
    ruleLayer = RULE_LAYER_PRETEXT_NATIVE,
    canvasPixelParityTarget = false,
    textDirection = corpus.textDirection,
    layoutEngine = layoutEngine,
    heightMetricSource = HEIGHT_METRIC_SOURCE_PLATFORM_TEXT_ENGINE_METRICS,
    fallbackReason = fallbackReason,
    driftKinds = collectDriftKinds(
      paragraph = paragraph,
      corpus = corpus,
      request = request,
      lineLayouts = lineLayouts,
      boundaryMap = boundaryMap,
    ).toTypedArray(),
    heightMetricDrivers = HEIGHT_METRIC_DRIVERS,
    breakTable = breakTable,
    boundaryMap = boundaryMap,
    complexShapeCounters = complexShapeCounters,
    lineDiagnostics = lineLayouts.map { line ->
      ParagraphLineDiagnostics(
        textStart = line.textStart.toDouble(),
        textEnd = line.textEnd.toDouble(),
        textDirection = corpus.textDirection,
        layoutEngine = line.layoutEngine,
        heightMetricSource = HEIGHT_METRIC_SOURCE_PLATFORM_TEXT_ENGINE_METRICS,
        fallbackReason = line.fallbackReason,
        driftKinds = collectLineDriftKinds(line, canonicalLayoutEngine).toTypedArray(),
        clusterViolationOffsets = collectLineClusterViolationOffsets(
          line,
          boundaryMap,
          paragraph.atomicSpans,
        ).map { it.toDouble() }.toDoubleArray(),
      )
    }.toTypedArray(),
  )
}

private fun buildPublicLayoutRequest(request: NativeLayoutRequest): ParagraphLayoutRequest {
  return ParagraphLayoutRequest(
    width = AndroidTextUnits.fromPx(request.width),
    left = AndroidTextUnits.fromPx(request.left),
    whiteSpace = request.whiteSpace,
    wordBreak = request.wordBreak,
    shapeSlices = request.shapeSlices.map { slice ->
      ParagraphShapeSlice(
        top = AndroidTextUnits.fromPx(slice.top),
        height = AndroidTextUnits.fromPx(slice.height),
        left = AndroidTextUnits.fromPx(slice.left),
        width = AndroidTextUnits.fromPx(slice.width),
      )
    }.toTypedArray(),
  )
}

private fun buildParagraphBreakTable(
  paragraph: NativePreparedParagraph,
  lineLayouts: List<NativeLineLayout>,
): ParagraphBreakTable {
  val hardBreaks = collectHardBreaks(paragraph.text)
  val hardBreakOffsets = hardBreaks.map { it.offset.toInt() }.toSet()
  val nativeSoftBreaks = lineLayouts
    .dropLast(1)
    .mapNotNull { line ->
      val offset = line.textEnd
      val nextCharIsHardBreak = paragraph.text.getOrNull(offset) == '\n'
      if (
        offset <= 0 ||
        offset >= paragraph.text.length ||
        hardBreakOffsets.contains(offset) ||
        nextCharIsHardBreak
      ) {
        null
      } else {
        ParagraphBreakOpportunity(
          offset = offset.toDouble(),
          kind = BREAK_KIND_NATIVE_SOFT,
          source = line.layoutEngine,
        )
      }
    }
  return ParagraphBreakTable(
    hardBreaks = hardBreaks.toTypedArray(),
    nativeSoftBreaks = nativeSoftBreaks.toTypedArray(),
    graphemeBoundaries = collectGraphemeBoundaries(paragraph).map { it.toDouble() }.toDoubleArray(),
    atomicSpans = paragraph.atomicSpans.map { span ->
      ParagraphAtomicSpan(
        textStart = span.start.toDouble(),
        textEnd = span.end.toDouble(),
        source = span.source,
      )
    }.toTypedArray(),
  )
}

private fun buildParagraphBoundaryMap(
  paragraph: NativePreparedParagraph,
  lineLayouts: List<NativeLineLayout>,
  breakTable: ParagraphBreakTable,
): ParagraphBoundaryMap {
  val graphemeBoundaries = breakTable.graphemeBoundaries.map { it.toInt() }.sorted()
  val runBoundaries = collectRunBoundaries(paragraph)
  val atomicSpanBoundaries = paragraph.atomicSpans
    .flatMap { span -> listOf(span.start, span.end) }
    .distinct()
    .sorted()
  val clusterViolations = collectParagraphClusterViolationOffsets(
    lineLayouts,
    graphemeBoundaries.toSet(),
    paragraph.atomicSpans,
  )

  return ParagraphBoundaryMap(
    utf16Length = paragraph.text.length.toDouble(),
    graphemeBoundaries = graphemeBoundaries.map { it.toDouble() }.toDoubleArray(),
    runBoundaries = runBoundaries.map { it.toDouble() }.toDoubleArray(),
    hardBreaks = breakTable.hardBreaks.map { it.offset }.toDoubleArray(),
    nativeSoftBreaks = breakTable.nativeSoftBreaks.map { it.offset }.toDoubleArray(),
    atomicSpanBoundaries = atomicSpanBoundaries.map { it.toDouble() }.toDoubleArray(),
    clusterViolationOffsets = clusterViolations.map { it.toDouble() }.toDoubleArray(),
  )
}

private fun collectHardBreaks(text: String): List<ParagraphBreakOpportunity> {
  val breaks = ArrayList<ParagraphBreakOpportunity>()
  var index = text.indexOf('\n')
  while (index >= 0) {
    breaks += ParagraphBreakOpportunity(
      offset = (index + 1).toDouble(),
      kind = BREAK_KIND_HARD,
      source = "source_newline",
    )
    index = text.indexOf('\n', index + 1)
  }
  return breaks
}

private fun collectGraphemeBoundaries(paragraph: NativePreparedParagraph): List<Int> {
  return collectGraphemeBoundariesForText(
    paragraph.text,
    paragraph.runs.firstOrNull()?.style?.locale ?: "",
  )
}

private fun collectRunBoundaries(paragraph: NativePreparedParagraph): List<Int> {
  return (listOf(0, paragraph.text.length) + paragraph.runs.flatMap { run ->
    listOf(run.start, run.end)
  })
    .map { it.coerceIn(0, paragraph.text.length) }
    .distinct()
    .sorted()
}

private fun collectParagraphClusterViolationOffsets(
  lineLayouts: List<NativeLineLayout>,
  graphemeBoundarySet: Set<Int>,
  atomicSpans: List<NativeAtomicSpan>,
): List<Int> {
  return lineLayouts
    .flatMap { line -> collectLineClusterViolationOffsets(line, graphemeBoundarySet, atomicSpans) }
    .distinct()
    .sorted()
}

private fun collectLineClusterViolationOffsets(
  line: NativeLineLayout,
  boundaryMap: ParagraphBoundaryMap,
  atomicSpans: List<NativeAtomicSpan>,
): List<Int> {
  return collectLineClusterViolationOffsets(
    line,
    boundaryMap.graphemeBoundaries.map { it.toInt() }.toSet(),
    atomicSpans,
  )
}

private fun collectLineClusterViolationOffsets(
  line: NativeLineLayout,
  graphemeBoundarySet: Set<Int>,
  atomicSpans: List<NativeAtomicSpan>,
): List<Int> {
  val violations = LinkedHashSet<Int>()
  if (!graphemeBoundarySet.contains(line.textStart)) {
    violations += line.textStart
  }
  if (!graphemeBoundarySet.contains(line.textEnd)) {
    violations += line.textEnd
  }
  atomicSpans.forEach { span ->
    if (line.textStart > span.start && line.textStart < span.end) {
      violations += line.textStart
    }
    if (line.textEnd > span.start && line.textEnd < span.end) {
      violations += line.textEnd
    }
  }
  return violations.toList()
}

private fun buildComplexShapeCounters(
  paragraph: NativePreparedParagraph,
  boundaryMap: ParagraphBoundaryMap,
): ParagraphComplexShapeCounters {
  val boundaries = boundaryMap.graphemeBoundaries.map { it.toInt() }
  return ParagraphComplexShapeCounters(
    bidiRunCount = countBidiRuns(paragraph.text).toDouble(),
    emojiClusterCount = countClusters(paragraph.text, boundaries, ::containsEmoji).toDouble(),
    complexClusterCount = countComplexClusters(paragraph.text, boundaries).toDouble(),
    clusterViolationCount = boundaryMap.clusterViolationOffsets.size.toDouble(),
  )
}

private fun countClusters(
  text: String,
  boundaries: List<Int>,
  predicate: (String) -> Boolean,
): Int {
  return boundaries
    .zipWithNext()
    .count { (start, end) -> end > start && predicate(text.substring(start, end)) }
}

private fun countComplexClusters(text: String, boundaries: List<Int>): Int {
  return boundaries
    .zipWithNext()
    .count { (start, end) ->
      if (end <= start) {
        false
      } else {
        val cluster = text.substring(start, end)
        cluster.length > Character.charCount(cluster.codePointAt(0)) ||
          cluster.codePoints().anyMatch { codePoint ->
            isCombiningMark(codePoint) ||
              isVariationSelector(codePoint) ||
              isEmojiModifier(codePoint) ||
              isRegionalIndicator(codePoint) ||
              isIndicVirama(codePoint) ||
              codePoint == ZERO_WIDTH_JOINER
          }
      }
    }
}

private fun countBidiRuns(text: String): Int {
  var runs = 0
  var previousDirection: Int? = null
  text.codePoints().forEachOrdered { codePoint ->
    val direction = when (Character.getDirectionality(codePoint)) {
      Character.DIRECTIONALITY_LEFT_TO_RIGHT -> 0
      Character.DIRECTIONALITY_RIGHT_TO_LEFT,
      Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC -> 1
      else -> null
    }
    if (direction != null && direction != previousDirection) {
      runs += 1
      previousDirection = direction
    }
  }
  return runs
}

private fun collectLineDriftKinds(
  line: NativeLineLayout,
  canonicalEngine: String,
): List<String> {
  val driftKinds = LinkedHashSet<String>()
  if (line.layoutEngine != canonicalEngine || line.fallbackReason != null) {
    driftKinds += DRIFT_ENGINE
  }
  if (line.fallbackReason != null) {
    driftKinds += DRIFT_HEIGHT_METRIC
  }
  return driftKinds.toList()
}

private fun collectDriftKinds(
  paragraph: NativePreparedParagraph,
  corpus: NativePreparedCorpus,
  request: NativeLayoutRequest,
  lineLayouts: List<NativeLineLayout>,
  boundaryMap: ParagraphBoundaryMap,
): List<String> {
  val driftKinds = LinkedHashSet<String>()
  val canonicalEngine = LAYOUT_ENGINE_ANDROID_STATIC_LAYOUT_COMPAT
  if (lineLayouts.any { it.layoutEngine != canonicalEngine || it.fallbackReason != null }) {
    driftKinds += DRIFT_ENGINE
  }
  if (
    request.whiteSpace != WHITE_SPACE_NORMAL ||
    request.wordBreak != WORD_BREAK_NORMAL ||
    request.shapeSlices.isNotEmpty() ||
    paragraph.atomicSpans.isNotEmpty()
  ) {
    driftKinds += DRIFT_ALGORITHM_RULE
    driftKinds += DRIFT_LINE_BREAK_STRATEGY
  }
  if (lineLayouts.any { it.fallbackReason != null }) {
    driftKinds += DRIFT_HEIGHT_METRIC
  }
  if (!corpus.includeFontPadding) {
    driftKinds += DRIFT_PADDING
  }
  if (paragraph.runs.any { it.style.locale.isNotBlank() }) {
    driftKinds += DRIFT_LOCALE_METRIC
  }
  if (containsPotentialFallbackGlyph(paragraph.text)) {
    driftKinds += DRIFT_FALLBACK_FONT
  }
  if (containsEmoji(paragraph.text)) {
    driftKinds += DRIFT_EMOJI_METRIC
  }
  if (boundaryMap.clusterViolationOffsets.isNotEmpty()) {
    driftKinds += DRIFT_CLUSTER_BOUNDARY
  }
  return driftKinds.toList()
}

private fun containsPotentialFallbackGlyph(text: String): Boolean {
  return text.codePoints().anyMatch { codePoint ->
    codePoint > 0x02AF
  }
}

private fun containsEmoji(text: String): Boolean {
  return text.codePoints().anyMatch { codePoint ->
    codePoint in 0x1F000..0x1FAFF ||
      codePoint in 0x2600..0x27BF ||
      codePoint == 0xFE0F
  }
}
